### PR TITLE
SCMOD-13301 Worker-Framework: Move away from Files.exists

### PR DESCRIPTION
**Pull request by simranjeetc**
https://portal.digitalsafe.net/browse/SCMOD-13301

### CODE REVIEW

File: worker-store-fs/src/main/java/com/hpe/caf/worker/datastore/fs/FileSystemDataStore.java [hunk 1]
```diff
@@ -79,7 +79,28 @@ public FileSystemDataStore(final FileSystemDataStoreConfiguration config)
 
         LOG.debug("Initialised");
     }
-
+    
+    private boolean doesFileExist(final Path path)
```
Comment (dermot-hardy) [inline worker-store-fs/src/main/java/com/hpe/caf/worker/datastore/fs/FileSystemDataStore.java]: Utility functions should be `static`.

**Review by dermot-hardy — COMMENTED** [review]
What you've done here is really debugging code for [SCMOD-13041](https://portal.digitalsafe.net/browse/SCMOD-13041) rather than the replacement implementation expected for [SCMOD-13301](https://portal.digitalsafe.net/browse/SCMOD-13301).  We can go ahead and check this in but make sure it is marked up as progressing SCMOD-13041.  For SCMOD-13301 the exception needs to be propagated so that the calling code can decide what to do.

File: worker-store-fs/src/main/java/com/hpe/caf/worker/datastore/fs/FileSystemDataStore.java [hunk 2]
```diff
@@ -79,7 +79,28 @@ public FileSystemDataStore(final FileSystemDataStoreConfiguration config)
...
+                LOG.error("file does not exist", ex);
+            }
+            else
+            {
+                LOG.error("File existence check failed ", ex);
+            }
+            return false;
+        }
+        return true;
+    }
```
Comment (dermot-hardy) [inline worker-store-fs/src/main/java/com/hpe/caf/worker/datastore/fs/FileSystemDataStore.java]: As as aside, this function would normally be formatted like this:

    private boolean doesFileExist(final Path path)
    {
        try {
            path.getFileSystem().provider().checkAccess(path);
        } catch (final IOException ex) {
            if (ex instanceof NoSuchFileException) {
                LOG.error("file does not exist", ex);
            } else {
                LOG.error("File existence check failed ", ex);
            }
            return false;
        }
        return true;
    }

File: worker-store-fs/src/main/java/com/hpe/caf/worker/datastore/fs/FileSystemDataStore.java [hunk 3]
```diff
@@ -65,11 +65,12 @@ public FileSystemDataStore(final FileSystemDataStoreConfiguration config)
         throws DataStoreException
     {
         dataStorePath = FileSystems.getDefault().getPath(config.getDataDir());
-        throwExceptionIfPathDoesNotExist(dataStorePath);
-        try {
-            Files.createDirectory(dataStorePath);
-        } catch (IOException e) {
-            throw new DataStoreException("Cannot create data store directory", e);
+        if (!Files.exists(dataStorePath)) {
```
Comment (simranjeetc) [inline worker-store-fs/src/main/java/com/hpe/caf/worker/datastore/fs/FileSystemDataStore.java]: @dermot-hardy Do we need to replace this as well? Here non-existence is expected whereas in places where we are replacing Files.exists, existence of path is expected. If we still want to replace this, then we may need to gobble up the exception and just log it. 

Let me know what you think.

**Review by dermot-hardy — CHANGES_REQUESTED** [review]
All usages of `Files.exists()` should be replaced.  This Jira is effectively banning the use of this method from our codebase because of its issues.

File: worker-store-fs/src/main/java/com/hpe/caf/worker/datastore/fs/FileSystemDataStore.java [hunk 4]
```diff
@@ -79,7 +79,27 @@ public FileSystemDataStore(final FileSystemDataStoreConfiguration config)
 
         LOG.debug("Initialised");
     }
-
+    
+    private static boolean pathDoesNotExist(final Path path)
```
Comment (dermot-hardy) [inline worker-store-fs/src/main/java/com/hpe/caf/worker/datastore/fs/FileSystemDataStore.java]: I must not be explaining this Jira correctly.  You've just replaced the `Files.exists` call with a method which has exactly the same problem.

**Review by dermot-hardy — CHANGES_REQUESTED** [review]

**Comment by dermot-hardy** [comment]
This is what a `doesFileExist()` method should look like:

    public static boolean doesFileExist(final Path path) throws IOException
    {
        try {
            path.getFileSystem().provider().checkAccess(path);
        } catch (final NoSuchFileException ex) {
            return false;
        }

        return true;
    }

Note that it only returns that the file does not exist if the exception when checking the access was `NoSuchFileException`.  Other `IOException`s are propagated.

File: worker-store-fs/src/main/java/com/hpe/caf/worker/datastore/fs/FileSystemDataStore.java (hunk 4 above)
Comment (simranjeetc) [reply worker-store-fs/src/main/java/com/hpe/caf/worker/datastore/fs/FileSystemDataStore.java]: Resolved with [this](https://github.com/WorkerFramework/worker-framework/pull/128#issuecomment-840929683) implementation.

**Review by dermot-hardy — APPROVED** [review]

**Pull request merged**